runtimes/ocaml/runtime.ml

```diff
@@ -45,42 +45,60 @@ type source_position = {
   law_headings : string list;
 }
 
-exception EmptyError
-exception AssertionFailed of source_position
-exception ConflictError of source_position
-exception UncomparableDurations
-exception IndivisibleDurations
-exception ImpossibleDate
-exception NoValueProvided of source_position
-exception NotSameLength
-exception Division_by_zero (* Shadows the stdlib definition *)
-
-(* Register exceptions printers *)
+type error =
+  | AssertionFailed
+  | NoValue
+  | Conflict
+  | DivisionByZero
+  | NotSameLength
+  | UncomparableDurations
+  | IndivisibleDurations
+
+let error_to_string = function
+  | AssertionFailed -> "AssertionFailed"
+  | NoValue -> "NoValue"
+  | Conflict -> "Conflict"
+  | DivisionByZero -> "DivisionByZero"
+  | NotSameLength -> "NotSameLength"
+  | UncomparableDurations -> "UncomparableDurations"
+  | IndivisibleDurations -> "IndivisibleDurations"
+
+let error_message = function
+  | AssertionFailed -> "this assertion doesn't hold"
+  | NoValue -> "no computation with valid conditions found"
+  | Conflict -> "two or more concurring valid computations"
+  | DivisionByZero -> "division by zero"
+  | NotSameLength -> "traversing multiple lists of different lengths"
+  | UncomparableDurations ->
+    "comparing durations in different units (e.g. months vs. days)"
+  | IndivisibleDurations -> "dividing durations that are not in days"
+
+exception Error of error * source_position list
+exception Empty
+
+let error err pos = raise (Error (err, pos))
+
+(* Register (fallback) exception printers *)
 let () =
-  let pos () p =
+  let ppos () p =
     Printf.sprintf "%s:%d.%d-%d.%d" p.filename p.start_line p.start_column
       p.end_line p.end_column
   in
-  let pr fmt = Printf.ksprintf (fun s -> Some s) fmt in
+  let pposl () pl = String.concat ", " (List.map (ppos ()) pl) in
   Printexc.register_printer
   @@ function
-  | EmptyError -> pr "A variable couldn't be resolved"
-  | AssertionFailed p -> pr "At %a: Assertion failed" pos p
-  | ConflictError p -> pr "At %a: Conflicting exceptions" pos p
-  | UncomparableDurations -> pr "Ambiguous comparison between durations"
-  | IndivisibleDurations -> pr "Ambiguous division between durations"
-  | ImpossibleDate -> pr "Invalid date"
-  | NoValueProvided p ->
-    pr "At %a: No definition applied to this variable" pos p
-  | NotSameLength -> pr "Attempt to traverse lists of different lengths"
-  | Division_by_zero -> pr "Division by zero"
+  | Error (err, pos) ->
+    Some (Printf.sprintf "At %a: %s" pposl pos (error_message err))
   | _ -> None
 
 let () =
   Printexc.set_uncaught_exception_handler
   @@ fun exc bt ->
   Printf.eprintf "[ERROR] %s\n%!" (Printexc.to_string exc);
   if Printexc.backtrace_status () then Printexc.print_raw_backtrace stderr bt
+(* TODO: the backtrace will point to the OCaml code; but we could make it point
+   to the Catala code if we add #line directives everywhere in the generated
```

Comment:
We could add those in the OCaml backend with the `-d` flag.

Reply:
Yes, I'll experiment with that a bit more: it's not perfect (granularity is the OCaml line of code; lots of backtrace items point to the runtime or libraries) and will make the OCaml code less readable but that can definitely be pretty useful for debugging.